pkg/cloudprovider/cloudprovider.go

```diff
@@ -168,3 +289,60 @@ func (c *CloudProvider) Name() string {
 func (c *CloudProvider) GetSupportedNodeClasses() []status.Object {
 	return []status.Object{&v1alpha1.GCENodeClass{}}
 }
+
+func (c *CloudProvider) instanceToNodeClaim(i *instance.Instance, instanceType *cloudprovider.InstanceType) *karpv1.NodeClaim {
+	nodeClaim := &karpv1.NodeClaim{}
+	labels := map[string]string{}
+	annotations := map[string]string{}
+
+	if instanceType != nil {
+		labels = utils.GetAllSingleValuedRequirementLabels(instanceType)
+
+		resourceFilter := func(name corev1.ResourceName, value resource.Quantity) bool {
+			return !resources.IsZero(value)
+		}
+
+		nodeClaim.Status.Capacity = lo.PickBy(instanceType.Capacity, resourceFilter)
+		nodeClaim.Status.Allocatable = lo.PickBy(instanceType.Allocatable(), resourceFilter)
+	}
+
+	// Set core labels
+	labels[corev1.LabelTopologyZone] = i.Location
+	labels[karpv1.CapacityTypeLabelKey] = i.CapacityType
+
+	// Add node pool label if present
+	if v, ok := i.Labels[karpv1.NodePoolLabelKey]; ok {
+		labels[karpv1.NodePoolLabelKey] = v
+	}
+
+	nodeClaim.Labels = labels
+	nodeClaim.Annotations = annotations
+	nodeClaim.CreationTimestamp = metav1.Time{Time: i.CreationTime}
+
+	// Handle instance deletion status (e.g., TERMINATED or STOPPING in GCP)
+	if i.Status == instance.InstanceStatusTerminated || i.Status == instance.InstanceStatusStopping {
+		nodeClaim.DeletionTimestamp = &metav1.Time{Time: time.Now()}
+	}
+
+	// ProviderID format for GCE: gce://project/zone/instance-name
+	nodeClaim.Status.ProviderID = fmt.Sprintf("gce://%s/%s/%s", i.ProjectID, i.Location, i.Name)
+	nodeClaim.Status.ImageID = i.ImageID
+
+	return nodeClaim
+}
+
+func (c *CloudProvider) resolveNodeClassFromNodeClaim(ctx context.Context, nodeClaim *karpv1.NodeClaim) (*v1alpha1.GCENodeClass, error) {
+	ref := nodeClaim.Spec.NodeClassRef
+	if ref == nil {
+		return nil, fmt.Errorf("nodeClaim missing NodeClassRef")
+	}
+	nodeClass := &v1alpha1.GCENodeClass{}
+	if err := c.kubeClient.Get(ctx, types.NamespacedName{Name: ref.Name}, nodeClass); err != nil {
+		return nil, err
+	}
+	return nodeClass, nil
+}
+
+func (c *CloudProvider) resolveInstanceTypes(ctx context.Context, nodeClass *v1alpha1.GCENodeClass) ([]*cloudprovider.InstanceType, error) {
```

Comment:
Does it make sense to have a separate helper function if it just calls `c.instanceTypeProvider.List` with same arguments and returns results as is?

Reply:
Thank you for your suggestion, but when instance type is nil, we can still set the karpenter label for nodeclaim to facilitate subsequent processing by karpenter, so we can keep the current implementation.

Reply:
Yes you are correct, but like other karpenter providers, we can provide some filtering methods in this method to enhance the available results after List.

https://github.com/cloudpilot-ai/karpenter-provider-alibabacloud/blob/a8e99d9f292d640ee0bce334d24240cb15501e7f/pkg/cloudprovider/cloudprovider.go#L341